File: 2690-house-robber-iv/house-robber-iv.py
def check(mid,nums,k):
    n=len(nums)
    cnt=0
    i=0
    while i<n:
        if nums[i]<=mid:
            cnt+=1
            i+=2
        else:
            i+=1
    return cnt>=k


class Solution:
    def minCapability(self, nums: List[int], k: int) -> int:
        l=min(nums)
        h=max(nums)
        ans=h
        while l<=h:
            mid=(l+h)//2
            if check(mid,nums,k):
                ans=mid
                h=mid-1
            else:
                l=mid+1
        return ans

        

File: 2690-house-robber-iv/house-robber-iv.cpp

class Solution {
public:
    int minCapability(vector<int>& nums, int k) {

        int l=*min_element(nums.begin(),nums.end());
        int h=*max_element(nums.begin(),nums.end());
        int ans=h;
        int n=nums.size();

        while(l<=h){
            int mid=(l+h)/2;
            int take=0;
            for(int i=0;i<n;i++){
                if(nums[i]<=mid){
                    take++;
                    i++;
                }
            }
            if(take>=k){
                ans=mid;
                h=mid-1;
            }
            else{
                l=mid+1;
            }
        }
        return ans;
    }
};